### ISSUE TITLE
SUBSCRIBE does not scale beyond 100s of subscribes -- high CRDB load may require an `ALTER SYSTEM` limit 

### ISSUE THREAD

**Issue by philip-stoev** [MaterializeInc/materialize#18261]
### What version of Materialize are you using?

e36ac08340a70009f13271c6d34c6fc5d688d9b4

### How did you install Materialize?

Docker image

### What is the issue?

If the `Subscribe` scenario in `test/limits` is attempted with 1000 SUBSCRIBEs, Mz will grind to a halt with very high CRDB and `clusterd` CPU usage. This seems to be the case regardless of whether the object being subscribed has an index or not.

To avoid a DDoS against CRDB, we need either:

- a fix that prevents the creation of a dataflow for each subscribe
or
- an `ALTER SYSTEM` limit that prevents too many subscribes from being created. Since there can only be one subscribe per connection, one option it to simply limit the total number of SQL connections allowed. This in turn will also limit the total number of concurrent `SELECTs` in the system, which would also be beneficial.

To reproduce:

```
diff --git a/test/limits/mzcompose.py b/test/limits/mzcompose.py
index e4b0bf108..e3ace8dd9 100644
--- a/test/limits/mzcompose.py
+++ b/test/limits/mzcompose.py
@@ -103,14 +103,13 @@ class Tables(Generator):
 
 
 class Subscribe(Generator):
-    COUNT = 100  # Each SUBSCRIBE instantiates a dataflow, so impossible to do 1K
-
     @classmethod
     def body(cls) -> None:
         print("> DROP TABLE IF EXISTS t1 CASCADE;")
         print("> CREATE TABLE t1 (f1 INTEGER);")
         print("> INSERT INTO t1 VALUES (-1);")
         print("> CREATE MATERIALIZED VIEW v1 AS SELECT COUNT(*) FROM t1;")
+        print("> CREATE DEFAULT INDEX ON v1;")
 
         for i in cls.all():
             print(
```

and then run:

```
cd test/limits
./mzcompose down -v
./mzcompose run default --scenario=Subscribe
```

**Comment by mjibson** [comment]
> This is expected with the high number of dataflows they need to maintain.

If there's an index, why are there so many dataflows it needs to maintain? Is it that each subscribe creates a sink dataflow, which there's no way around?

**Comment by benesch** [comment]
> > is is expected with the high number of dataflows they need to maintain.
> 
> If there's an index, why are there so many dataflows it needs to maintain? Is it that each subscribe creates a sink dataflow, which there's no way around?

Precisely. We might hope to optimize this by somehow reusing existing dataflows when an identical subscribe comes in, but at present, we do not.

**Comment by chaas** [comment]
Filed this https://github.com/MaterializeInc/materialize/issues/18341 for adding limits to subscribes/sql connections per region.

@teskje we can hand this off to compute to track to a potential optimization to prevent the creation of a dataflow for each subscribe

**Comment by teskje** [comment]
> we can hand this off to compute to track to a potential optimization to prevent the creation of a dataflow for each subscribe

Sounds good, though we should probably also check where the 120% CRDB load is coming from. Seems a bit much for a single materialized view.

**Comment by teskje** [comment]
I just tested running 1000 subscribes like the ones in the limits test case on staging on an xsmall instance and it barely broke a sweat (the replica is green, yellow is environmentd):

![Image](https://user-images.githubusercontent.com/4521314/234574715-7a8c6be7-e9af-4d34-b2a1-dd096372e114.png)

So it seems like reducing the number of dataflows wouldn't actually be very helpful, at least not for reducing CPU usage (certainly for reducing CRDB reads when there is no index).

It's worth pointing out that in my test I simply opened a bunch of long-running psql sessions that all ran `SUBSCRIBE v1`. This is different from the test case, which instantiates cursors and then runs `FETCH ALL` on them once. I think that means in the test case we don't have long running subscribes and instead the subscribes are torn down immediately after they have produced data. Is it possible that the CPU usage we see there isn't caused by running the dataflows but rather by creating them?
